geo/src/algorithm/validation/polygon.rs

```diff
@@ -153,10 +196,212 @@ impl<F: GeoFloat> Validation for Polygon<F> {
+        // Build PreparedGeometry for the whole polygon here (deferred from earlier)
+        // so we don't pay the cost if validation already failed above.
+        let prepared_polygon = PreparedGeometry::from(self);
+        if let Some(problem_coords) =
+            check_interior_simply_connected_from_graph(&prepared_polygon.geometry_graph)
+        {
+            handle_validation_error(InvalidPolygon::InteriorNotSimplyConnected(problem_coords))?;
+        }
+
         Ok(())
     }
 }
 
+/// Check that the polygon interior is simply connected using the GeometryGraph.
+///
+/// This function extracts touch point information from a pre-computed GeometryGraph,
+/// which is more efficient than brute-force comparison because the graph's R-tree
+/// intersection detection runs in O((n + k) log n) instead of O(V²).
+///
+/// The interior is disconnected if rings touch in a way that creates separate
+/// interior regions. This occurs when:
+/// - Two rings share 2+ touch points at different coordinates
+/// - Rings form a cycle through distinct single touch points
+///
+/// Note: Multiple rings meeting at a single coordinate does NOT disconnect
+/// the interior. For example, three triangular holes meeting at one point
+/// can still have a connected interior around their perimeter.
+///
+/// Returns `None` if the interior is simply connected, or `Some(coords)` with
+/// the problematic touch coordinates if the interior is disconnected.
+fn check_interior_simply_connected_from_graph<F: GeoFloat>(
+    graph: &GeometryGraph<F>,
+) -> Option<Vec<Coord<f64>>> {
+    let edges = graph.edges();
+
+    // Need at least 2 edges (rings) to have a touch point
+    // Edge 0 = exterior, Edge 1+ = interiors
+    if edges.len() < 2 {
+        return None;
+    }
+
+    // Build ring-pair touch map from edge intersections
+    // Two edges that meet at a point will both have an EdgeIntersection at that coordinate
+    let mut ring_pair_touch_coords: HashMap<(usize, usize), HashSet<(u64, u64)>> = HashMap::new();
+
+    // Collect vertex coordinates from each edge for filtering
+    // We only care about intersections where at least one ring has the point as a vertex
+    // (i.e., touches), not crossing intersections in the middle of both segments
+    let edge_vertices: Vec<HashSet<(u64, u64)>> = edges
+        .iter()
+        .map(|edge: &std::rc::Rc<RefCell<Edge<F>>>| {
+            let edge = RefCell::borrow(edge);
+            edge.coords()
+                .iter()
+                .filter_map(|c| coord_to_bits(c))
+                .collect()
+        })
+        .collect();
+
+    // Collect all intersection coordinates from each edge
+    let edge_intersections: Vec<Vec<(u64, u64)>> = edges
+        .iter()
+        .map(|edge: &std::rc::Rc<RefCell<Edge<F>>>| {
+            let edge = RefCell::borrow(edge);
+            edge.edge_intersections()
+                .iter()
+                .filter_map(|ei| coord_to_bits(&ei.coordinate()))
```

Comment:
re: `f64::to_bits`

Is this in pursuit of getting the floating points to fit into a HashSet? There's a reason you can't put floats in a HashSet, and it's not clear to me that this usage is valid.

Can you talk more about this?